10up/10up-toolkit
# HMR and Fast Refresh

### Description of the Change




Closes #85 


This PR introduces React Fast Refresh for block development and Hot Module Reloading for front-end css.

![react-fast-refresh-toolkit](https://user-images.githubusercontent.com/6104632/155181035-b77a53f8-6a45-454d-934c-5667bbb0f06a.gif)
### Alternate Designs

Refresh the browser every time you make a change to a block.

### Possible Drawbacks

Not every code change can be reloaded without a refresh. Based on my quick tests, changing `registerBlockType` args causes a full page reload. However, for most of the block development, fast refresh should work fine.



This change is fully backward compatible and is only effective if you pass the `hot` option.

~Unfortunately, BrowserSync is still required as this setup only works for block development. Hopefully, in the future we can get rid of BrowserSync in favor of HMR.~

We can now deprecate BrowserSync.

### Verification Process Option 1 (Installing from npm)

Docs are also available here: https://github.com/10up/10up-toolkit/tree/feature/fast-refresh/packages/toolkit#fast-refresh

- Install `10up-toolkit@2.2.0-alpha.4` via npm.
- Run `10up-toolkit watch --hot` or if using the wp-scaffold `npm run watch -- --hot`
- Set `SCRIPT_DEBUG` to `true`
- Open up `functions.php` or the entry point of your plugin and add the following snippet of code (replace the constants with the URL and PATH to dist folder):
  - https://github.com/10up/10up-toolkit/blob/feature/fast-refresh/projects/10up-theme/functions.php#L17
```php
if ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG && file_exists( __DIR__ . '/dist/fast-refresh.php' ) ) {
	define( 'TENUP_TOOLKIT_DIST_URL', TENUP_THEME_DIST_URL );
	define( 'TENUP_TOOLKIT_DIST_PATH', TENUP_THEME_DIST_PATH );
	require_once __DIR__ . '/dist/fast-refresh.php';
}
```
Make sure to reload the page after running 10up-toolkit as the `dist/fast-refresh.php` file is generated by 10up-toolkit

### Verification Process Option 2 (using this repo)
- This has only been tested with Chrome! Let me know if it works on other browsers.
- Clone this repo and switch to this branch
  - `git clone git@github.com:10up/10up-toolkit.git && cd 10up-toolkit && git checkout feature-fast-refresh`
  - From the root dir, run `npm install`
- Create or use an existing WordPress environment form where you can run `projects/10up-theme`. If you are using `wp-local-docker` you can map the path to `projects/10up-theme` to your containers

Nginx container:
```yaml
  nginx:
    .....
    volumes:
      - './wordpress:/var/www/html:cached'
      - '~/projects/10up-toolkit/projects/10up-theme:/var/www/html/wp-content/themes/10up-theme'
      - './config/nginx/default.conf:/etc/nginx/conf.d/default.conf:cached'
```

phpfpm container:
```yaml
phpfpm:
    ....
    volumes:
      - './wordpress:/var/www/html:cached'
      - '~/projects/10up-toolkit/projects/10up-theme:/var/www/html/wp-content/themes/10up-theme'
      - './config/php-fpm/docker-php-ext-xdebug.ini:/etc/php.d/docker-php-ext-xdebug.ini:cached'
      - 'wplocaldockerCache:/var/www/.wp-cli/cache:cached'
      - '~/.ssh:/home/nicholas_io/.ssh:cached'
      - './config/php-fpm/wp-cli.local.yml:/var/www/.wp-cli/config.yml:cached'
```

Replace `~/projects/10up-toolkit/projects/10up-theme` with the path to your local version of `10up-toolkit/projects/10up-theme`
- Set `SCRIPT_DEBUG` to true in `wp-config.php`
- Go to Appearance -> Themes and enable the "10up theme"
- Go to `10up-toolkit/projects/10up-theme` and run `npm run watch -- --hot`
- Go to any post or page and add the "Example Block"
- Try making changes to in `includes/blocks/example-block/edit.js` and `includes/blocks/example-block.css` you should see changes in real-time.
  -  The first "reload" is a little slower.

NOTE: if you have the Gutenberg plugin enabled you might see an error about service worker registration, just close that overlay and keep testing.
### Checklist:






### Changelog Entry



Added - React Fast Refresh for block development (BETA)
### Credits

This was heavily based on https://github.com/WordPress/gutenberg/pull/28273

The approach here was extended a bit and 10up-toolkit does not actually require running the Gutenberg plugin.


Props @
